src/ImageCore.jl

```diff
@@ -30,6 +29,16 @@ Color1Array{C<:Color1,N} = AbstractArray{C,N}
 const RRArray{To,From,N,M,P} = Base.ReshapedArray{To,N,Base.ReinterpretArray{To,M,From,P}}
 const RGArray = Union{Base.ReinterpretArray{<:AbstractGray,M,<:Number,P}, Base.ReinterpretArray{<:Number,M,<:AbstractGray,P}} where {M,P}
 
+# delibrately not export these constants to enable extensibility for downstream packages
+const NumberLike{T<:Number} = Union{T,AbstractGray{T}}
+const RealLike{T<:Real} = NumberLike{T}
+const FloatLike{T<:AbstractFloat} = RealLike{T}
+const FractionalLike{T<:Union{FixedPoint, AbstractFloat}} = RealLike{T}
+const IntegerLike{T<:Union{FixedPoint, Integer}} = RealLike{T}
```

Comment:
Where does this come up? (What applications does it have?)

Reply:
There might be some functions support input of both kinds

* `FixedPoint`
* `Gray{FixedPoint}`
* `Integer` (as an additional support)

Reply:
Let's add this when you know you need it. Integer values are problematic because so many people have the equation `255 == 1.0 == white` in their heads due to other languages not being able to do the kind of disambiguation that Julia can do. Try `RGB(255, 0, 128)` vs `RGB(255.0, 0.0, 128.0)` to see some of the guards we put in place for this. (We allow the latter because mathematics on images routinely returns values that are out-of-gamut, e.g., if you want to compute the mean color of an image you'd first sum all the pixels and get some huge values, but at the end you divide by the number of pixels and this puts you back in-gamut.)

Reply:
> so many people have the equation `255 == 1.0 == white` in their heads due to other languages not being able to do the kind of disambiguation that Julia can do

I'm still trying to get rid of this eq 😄 So a rule of thumb is to avoid `Integer` as possible as we can?

Reply:
Yes. Integers bad bad :smile:.

I agree, it's a totally pervasive issue. You don't really *feel* how horrible it is until you start trying to write code that works on a wide variety of input types and interacts with visualization that has to make reasonable default choices. As bad as 255 == 1.0 is, it's even worse if you have to use 4095 == 255 == 1.0 :smile: which is the appropriate equation if you're analyzing images collected with a 12-bit camera. It is revealing that explaining this pitfall is basically the first thing that the [scikit images docs](https://scikit-image.org/docs/dev/user_guide/data_types.html) do.

Reply:
I've been thinking ColorTypes needs a logo. I propose

![warning](https://user-images.githubusercontent.com/1525481/56864099-162a9600-6984-11e9-9583-6dd6e0b7ced9.png)

:stuck_out_tongue: